pyads/symbol.py

```diff
@@ -0,0 +1,277 @@
+"""
+Define the Symbol class
+
+Separate file because it depends on many other files, so we try to simplify
+the circular dependencies.
+"""
+
+from typing import TYPE_CHECKING, Any, Union, Optional, Type, List, Tuple, \
+    Callable
+# ads.Connection relies on structs.AdsSymbol (but in type hints only), so use
+# this 'if' to only include it when type hinting (False during execution)
+if TYPE_CHECKING:
+    from .ads import Connection  # pragma: no cover
+import re
+from ctypes import sizeof
+
+from .pyads_ex import adsGetSymbolInfo
+from .structs import NotificationAttrib
+from . import constants  # To access all constants, use package notation
+
+
+class AdsSymbol:
+    """Object that points to an ADS variable
+
+    Contains index group, index offset, name, symbol type, comment of ADS
+    symbol. Also remembers a reference to a Connection to be able to
+    read/write directly.
+
+    The virtual property `value` can be used to read from and write to
+    the symbol.
+
+    :ivar index_group: Index group of symbol
+    :ivar index_offset: Index offset of symbol
+    :ivar name: Name of symbol
+    :ivar type_name: String representation of symbol type (PLC-style,
+                     e.g. "LREAL")
+    :ivar _type_hint: Object from which the symtype will be deduced
+    :ivar plc_type: ctypes type of variable (from constants.PLCTYPE_*)
+    :ivar comment: Comment of symbol
+    """
+
+    def __init__(self,
+                 plc: "Connection",
+                 name: Optional[str] = None,
+                 index_group: Optional[int] = None,
+                 index_offset: Optional[int] = None,
+                 type_hint: Optional[Union[Type, str, int]] = None,
```

Comment:
Can we not just use type_hint: Optional[str] here? AdsSymbol objects are only created in Connection.get_symbol and Connection.get_all_symbols. The first one doesn't provide type_hint and the second passes a string. So I would rename type_hint to symbol_type and make it Optional[str]. This also makes the following logic easier.

Reply:
Yeah, let's do that.

At first I was using `info.dataType` (an integer in SymbolInfo) to resolve the type, since resolving an integer would be faster and more robust than matching a string.
Then I noticed this fails for arrays. But let's go all the way then.